File: agents/monitoring/domain/event.py
"""
MonitoringEvent - 监控事件领域模型

不可变值对象，用于表示监控系统中的各种事件。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum, auto
from typing import Any, Optional
from uuid import UUID, uuid4


class EventPriority(Enum):
    """事件优先级"""
    CRITICAL = 0   # 系统关键事件：错误、状态变更
    HIGH = 1       # 用户关注事件：消息完成、工具结果
    NORMAL = 2     # 常规事件：token 流、工具调用
    LOW = 3        # 诊断事件：性能指标、调试信息


class EventType(Enum):
    """事件类型枚举"""
    # === Agent 生命周期 ===
    AGENT_STARTED = "agent:started"
    AGENT_STOPPED = "agent:stopped"
    AGENT_ERROR = "agent:error"
    AGENT_PAUSED = "agent:paused"
    AGENT_RESUMED = "agent:resumed"

    # === 消息流事件 ===
    MESSAGE_START = "message:start"
    MESSAGE_DELTA = "message:delta"
    MESSAGE_COMPLETE = "message:complete"
    REASONING_DELTA = "reasoning:delta"

    # === 工具调用事件 ===
    TOOL_CALL_START = "tool_call:start"
    TOOL_CALL_END = "tool_call:end"
    TOOL_CALL_ERROR = "tool_call:error"
    TOOL_RESULT = "tool:result"

    # === 子智能体事件 ===
    SUBAGENT_SPAWNED = "subagent:spawned"
    SUBAGENT_STARTED = "subagent:started"
    SUBAGENT_PROGRESS = "subagent:progress"
    SUBAGENT_COMPLETED = "subagent:completed"
    SUBAGENT_FAILED = "subagent:failed"

    # === 后台任务事件 ===
    BG_TASK_QUEUED = "bg_task:queued"
    BG_TASK_STARTED = "bg_task:started"
    BG_TASK_PROGRESS = "bg_task:progress"
    BG_TASK_COMPLETED = "bg_task:completed"
    BG_TASK_FAILED = "bg_task:failed"

    # === 状态机事件 ===
    STATE_TRANSITION = "state:transition"
    STATE_ENTER = "state:enter"
    STATE_EXIT = "state:exit"

    # === 资源使用事件 ===
    TOKEN_USAGE = "metrics:tokens"
    MEMORY_USAGE = "metrics:memory"
    LATENCY_METRIC = "metrics:latency"

    # === Todo 管理事件 ===
    TODO_CREATED = "todo:created"
    TODO_UPDATED = "todo:updated"
    TODO_COMPLETED = "todo:completed"


@dataclass(frozen=True)
class MonitoringEvent:
    """
    监控事件值对象

    特性:
    - 不可变性 (frozen=True)
    - 唯一标识 (UUID)
    - 层级关系 (parent_id)
    - 优先级支持
    - 完整序列化

    Example:
        >>> event = MonitoringEvent(
        ...     type=EventType.AGENT_STARTED,
        ...     dialog_id="dialog-123",
        ...     source="SFullAgent",
        ...     context_id=uuid4(),
        ...     payload={"agent_name": "TeamLeadAgent"}
        ... )
        >>> child = MonitoringEvent.create_child(
        ...     parent=event,
        ...     type=EventType.SUBAGENT_SPAWNED,
        ...     payload={"subagent_name": "ExploreAgent"}
        ... )
    """

    type: EventType
    dialog_id: str
    source: str
    context_id: UUID
    timestamp: datetime = field(default_factory=datetime.utcnow)
    id: UUID = field(default_factory=uuid4)
    parent_id: Optional[UUID] = None
    priority: EventPriority = EventPriority.NORMAL
    payload: dict[str, Any] = field(default_factory=dict)
    metadata: dict[str, Any] = field(default_factory=dict)

    def is_child_of(self, parent: MonitoringEvent) -> bool:
        """
        检查是否为指定事件的子事件

        Args:
            parent: 可能的父事件

        Returns:
            True 如果当前事件是 parent 的子事件
        """
        return self.parent_id == parent.id

    def get_duration_ms(self, since: datetime) -> int:
        """
        计算自指定时间以来的毫秒数

        Args:
            since: 起始时间

        Returns:
            毫秒数
        """
        return int((self.timestamp - since).total_seconds() * 1000)

    def to_dict(self) -> dict[str, Any]:
        """
        序列化为字典

        Returns:
            可序列化的字典
        """
        return {
            "id": str(self.id),
            "type": self.type.value,
            "dialog_id": self.dialog_id,
            "timestamp": self.timestamp.isoformat(),
            "source": self.source,
            "context_id": str(self.context_id),
            "parent_id": str(self.parent_id) if self.parent_id else None,
            "priority": self.priority.value,
            "payload": self.payload,
            "metadata": self.metadata,
        }

    @staticmethod
    def create_child(
        parent: MonitoringEvent,
        type: EventType,
        payload: dict[str, Any],
        priority: EventPriority = EventPriority.NORMAL,
        metadata: Optional[dict[str, Any]] = None
    ) -> MonitoringEvent:
        """
        工厂方法：创建子事件

        自动继承父事件的 dialog_id, source, context_id
        设置 parent_id 为父事件的 id

        Args:
            parent: 父事件
            type: 子事件类型
            payload: 子事件载荷
            priority: 优先级（默认 NORMAL）
            metadata: 元数据（默认空字典）

        Returns:
            新的子事件实例
        """
        return MonitoringEvent(
            type=type,
            dialog_id=parent.dialog_id,
            source=parent.source,
            context_id=parent.context_id,
            parent_id=parent.id,
            priority=priority,
            payload=payload,
            metadata=metadata or {}
        )

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> MonitoringEvent:
        """
        从字典反序列化

        Args:
            data: 序列化的字典

        Returns:
            MonitoringEvent 实例
        """
        return cls(
            id=UUID(data["id"]) if "id" in data else uuid4(),
            type=EventType(data["type"]),
            dialog_id=data["dialog_id"],
            timestamp=datetime.fromisoformat(data["timestamp"]) if "timestamp" in data else datetime.utcnow(),
            source=data["source"],
            context_id=UUID(data["context_id"]),
            parent_id=UUID(data["parent_id"]) if data.get("parent_id") else None,
            priority=EventPriority(data["priority"]) if "priority" in data else EventPriority.NORMAL,
            payload=data.get("payload", {}),
            metadata=data.get("metadata", {})
        )

File: web/src/monitoring/domain/Event.ts
/**
 * MonitoringEvent - 监控事件领域模型
 *
 * 与后端 MonitoringEvent 对应，不可变值对象
 */

export enum EventPriority {
  CRITICAL = 0,
  HIGH = 1,
  NORMAL = 2,
  LOW = 3,
}

export enum EventType {
  // Agent 生命周期
  AGENT_STARTED = 'agent:started',
  AGENT_STOPPED = 'agent:stopped',
  AGENT_ERROR = 'agent:error',
  AGENT_PAUSED = 'agent:paused',
  AGENT_RESUMED = 'agent:resumed',

  // 消息流
  MESSAGE_START = 'message:start',
  MESSAGE_DELTA = 'message:delta',
  MESSAGE_COMPLETE = 'message:complete',
  REASONING_DELTA = 'reasoning:delta',

  // 工具调用
  TOOL_CALL_START = 'tool_call:start',
  TOOL_CALL_END = 'tool_call:end',
  TOOL_CALL_ERROR = 'tool_call:error',
  TOOL_RESULT = 'tool:result',

  // 子智能体
  SUBAGENT_SPAWNED = 'subagent:spawned',
  SUBAGENT_STARTED = 'subagent:started',
  SUBAGENT_PROGRESS = 'subagent:progress',
  SUBAGENT_COMPLETED = 'subagent:completed',
  SUBAGENT_FAILED = 'subagent:failed',

  // 后台任务
  BG_TASK_QUEUED = 'bg_task:queued',
  BG_TASK_STARTED = 'bg_task:started',
  BG_TASK_PROGRESS = 'bg_task:progress',
  BG_TASK_COMPLETED = 'bg_task:completed',
  BG_TASK_FAILED = 'bg_task:failed',

  // 状态机
  STATE_TRANSITION = 'state:transition',
  STATE_ENTER = 'state:enter',
  STATE_EXIT = 'state:exit',

  // 资源使用
  TOKEN_USAGE = 'metrics:tokens',
  MEMORY_USAGE = 'metrics:memory',
  LATENCY_METRIC = 'metrics:latency',

  // Todo
  TODO_CREATED = 'todo:created',
  TODO_UPDATED = 'todo:updated',
  TODO_COMPLETED = 'todo:completed',
}

export interface MonitoringEventData {
  id?: string;
  type: EventType;
  timestamp?: string;
  source: string;
  dialogId: string;
  contextId: string;
  parentId?: string;
  priority?: EventPriority;
  payload?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export class MonitoringEvent {
  private readonly _id: string;
  private readonly _type: EventType;
  private readonly _timestamp: Date;
  private readonly _source: string;
  private readonly _dialogId: string;
  private readonly _contextId: string;
  private readonly _parentId?: string;
  private readonly _priority: EventPriority;
  private readonly _payload: Record<string, unknown>;
  private readonly _metadata: Record<string, unknown>;

  constructor(params: MonitoringEventData) {
    this._id = params.id ?? crypto.randomUUID();
    this._type = params.type;
    this._timestamp = params.timestamp ? new Date(params.timestamp) : new Date();
    this._source = params.source;
    this._dialogId = params.dialogId;
    this._contextId = params.contextId;
    this._parentId = params.parentId;
    this._priority = params.priority ?? EventPriority.NORMAL;
    this._payload = params.payload ?? {};
    this._metadata = params.metadata ?? {};
  }

  // Getters
  get id(): string {
    return this._id;
  }

  get type(): EventType {
    return this._type;
  }

  get timestamp(): Date {
    return this._timestamp;
  }

  get source(): string {
    return this._source;
  }

  get dialogId(): string {
    return this._dialogId;
  }

  get contextId(): string {
    return this._contextId;
  }

  get parentId(): string | undefined {
    return this._parentId;
  }

  get priority(): EventPriority {
    return this._priority;
  }

  get payload(): Record<string, unknown> {
    return { ...this._payload };
  }

  get metadata(): Record<string, unknown> {
    return { ...this._metadata };
  }

  /**
   * 检查是否为指定事件的子事件
   */
  isChildOf(parent: MonitoringEvent): boolean {
    return this._parentId === parent.id;
  }

  /**
   * 计算自指定时间以来的毫秒数
   */
  getDurationMs(since: Date): number {
    return this._timestamp.getTime() - since.getTime();
  }

  /**
   * 序列化为 JSON
   */
  toJSON(): object {
    return {
      id: this._id,
      type: this._type,
      timestamp: this._timestamp.toISOString(),
      source: this._source,
      dialogId: this._dialogId,
      contextId: this._contextId,
      parentId: this._parentId,
      priority: this._priority,
      payload: this._payload,
      metadata: this._metadata,
    };
  }

  /**
   * 从 WebSocket 数据创建事件
   *
   * 后端会添加 `monitor:` 前缀到所有事件类型，这里需要移除前缀
   * 例如：`monitor:agent:started` -> `agent:started`
   */
  static fromWebSocket(data: unknown): MonitoringEvent {
    const parsed = data as Record<string, unknown>;

    // 移除 monitor: 前缀
    let eventTypeStr = parsed.type as string;
    if (eventTypeStr.startsWith('monitor:')) {
      eventTypeStr = eventTypeStr.slice(8); // 移除 'monitor:' 前缀
    }

    // 将字符串转换为 EventType 枚举值
    // 首先尝试直接匹配枚举值
    let eventType: EventType | undefined;

    // 遍历所有 EventType 枚举值，找到匹配的
    for (const [key, value] of Object.entries(EventType)) {
      if (value === eventTypeStr) {
        eventType = value as EventType;
        break;
      }
    }

    // 如果没找到匹配，使用字符串值（兼容未知事件类型）
    if (!eventType) {
      console.warn(`[MonitoringEvent] Unknown event type: ${eventTypeStr}`);
      eventType = eventTypeStr as EventType;
    }

    return new MonitoringEvent({
      id: parsed.id as string,
      type: eventType,
      timestamp: parsed.timestamp as string,
      source: parsed.source as string,
      dialogId: parsed.dialog_id as string,
      contextId: parsed.context_id as string,
      parentId: parsed.parent_id as string | undefined,
      priority: parsed.priority as EventPriority | undefined,
      payload: parsed.payload as Record<string, unknown> | undefined,
      metadata: parsed.metadata as Record<string, unknown> | undefined,
    });
  }

  /**
   * 创建子事件
   */
  static createChild(
    parent: MonitoringEvent,
    type: EventType,
    payload: Record<string, unknown>,
    priority: EventPriority = EventPriority.NORMAL,
    metadata?: Record<string, unknown>
  ): MonitoringEvent {
    return new MonitoringEvent({
      type,
      source: parent.source,
      dialogId: parent.dialogId,
      contextId: parent.contextId,
      parentId: parent.id,
      priority,
      payload,
      metadata,
    });
  }
}
